### PR TITLE
[Checkout] Be able to modify all past steps of the checkout

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

File: features/checkout/being_able_to_modify_remaining_steps.feature [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "PHP T-Shirt" priced at "$19.99"
+        And the store ships everywhere for free
+        And the store allows paying offline
+        And I am a logged in customer
+
+    @todo
+    Scenario: Placing an order after navigating to to the past steps of a checkout
```
Comment (michalmarcinkowski) [inline features/checkout/being_able_to_modify_remaining_steps.feature]: Scenarios should be more specific: `Placing an order after moving back from the checkout summary to the addressing step but without any address modification`

File: features/checkout/being_able_to_modify_remaining_steps.feature [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        And I confirm my order
+        Then I should be notified that the order has been successfully placed
+
+    @todo
+    Scenario: Addressing my order after selecting payment method
+        Given I specified the shipping address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        And I have proceeded order with "Free" shipping method and "Offline" payment
+        When I go back to the addressing step of the checkout
+        And I change the shipping address to „XX"
+        And I complete addressing step
```
Comment (michalmarcinkowski) [inline features/checkout/being_able_to_modify_remaining_steps.feature]: Probably can be squashed into step similar to `And I proceed specifying "XX" address`

**Comment by gabiudrescu** [comment]
I have a question about your workflow: I've seen a couple of PRs with scenarios only - no context, no other code written.

do they address some functionality that does not have testing or are they build before the development is done, as a part of your BDD / TDD approach?

I'm asking this because the functionality described here is something I'd also want in the project I'm working on and I don't know if I should write a story like "backport this from dev-master to our 0.18 version" or not.

**Comment by Arminek** [comment]
@gabiudrescu this scenario have @todo tag so it is ignored in test scope.

**Comment by michalmarcinkowski** [comment]
@gabiudrescu the PRs with scenarios only is a part of our workflow (BDD). Based on examples we can ensure the developer implementing specific feature has a good understanding of it. Product Owner can correct potential misunderstandings of the business logic even before any code is written. Providing feedback early can save a lot of time that is often wasted on implementation of something that does not work exactly as we expect and needs reworking.

Sometimes we merge PRs without any code, but with scenarios only because some features are already implemented and working exactly as we expect so there is no need to modify the implementation. Talking about this specific feature, it is not implemented yet (as @Arminek mentioned - "todo" tag), but for sure will be merged to the master branch in 2 weeks :)

**Comment by gabiudrescu** [comment]
thank you all for the details. now it's clear for me. 

I'll keep an eye on these PRs and offer you feedback - as they're my area of expertise more than is code.

File: features/checkout/being_able_to_modify_remaining_steps.feature [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        And the store has a product "PHP T-Shirt" priced at "$19.99"
+        And the store ships everywhere for free
+        And the store has "Raven Post" shipping method with "$10.00" fee
+        And the store allows paying offline
+        And I am a logged in customer
+
+    @todo
+    Scenario: Placing an order after moving back from the checkout summary to the addressing step but without any address modification
+        Given I am at the checkout summary step
+        When I go back to addressing step of the checkout
```
Comment (tuka217) [inline features/checkout/being_able_to_modify_remaining_steps.feature]: Couldn't it be "checkout addressing step"? Then we have "be on the checkout X step" and "back to the checkout X step".

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 4]
```diff
@@ -428,6 +447,22 @@ public function iProceedSelectingPaymentMethod($paymentMethodName = 'Offline')
     }
 
     /**
+     * @When I do not modify shipping method
+     */
+    public function iDoNotModifyShippingMethod()
+    {
+        $this->selectShippingPage->selectShippingMethod(null);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: I think we shouldn't even set `null` in this step. Can't we leave this method empty?

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 5]
```diff
@@ -463,6 +498,14 @@ public function iProceedLoggingAsGuestWithAsShippingCountry($email, CountryInter
     }
 
     /**
+     * @Given I return to the checkout summary step
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: **@When**

File: features/checkout/being_able_to_modify_remaining_steps.feature [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        And the store has "Raven Post" shipping method with "$10.00" fee
+        And the store allows paying offline
+        And I am a logged in customer
+
+    @ui
+    Scenario: Placing an order after moving back from the checkout summary to the addressing step but without any address modification
+        Given I had product "PHP T-Shirt" in the cart
+        And I was at the checkout summary step
+        When I go back to addressing step of the checkout
+        But I do not modify address
```
Comment (michalmarcinkowski) [inline features/checkout/being_able_to_modify_remaining_steps.feature]: How about more generic step `But I do not modify anything`?

File: features/checkout/being_able_to_modify_remaining_steps.feature [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        And I specify the shipping address as "Ankh Morpork", "Fire Alley", "90350", "United States" for "Jon Snow"
+        And I complete the addressing step
+        Then I should be on the checkout shipping step
+
+    @ui
+    Scenario: Addressing my order after selecting shipping method
+        Given I had product "PHP T-Shirt" in the cart
+        And I specified the shipping address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        And I select "Free" shipping method
+        And I complete the shipping step
```
Comment (michalmarcinkowski) [inline features/checkout/being_able_to_modify_remaining_steps.feature]: Can be squashed into `And I proceeded selecting "Free" shipping method`

File: features/checkout/being_able_to_modify_remaining_steps.feature [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        When I go back to addressing step of the checkout
+        And I specify the shipping address as "Ankh Morpork", "Fire Alley", "90350", "United States" for "Jon Snow"
```
Comment (michalmarcinkowski) [inline features/checkout/being_able_to_modify_remaining_steps.feature]: `I specify the shipping address as` -> `I change shipping address to`

**Comment by michalmarcinkowski** [comment]
We're missing scenarios for e.g. readdressing, addressed order:
```
       Given I had product "PHP T-Shirt" in the cart
       And I specified the shipping address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
        When I go back to addressing step of the checkout
        And I change shipping address to "Ankh Morpork", "Fire Alley", "90350", "United States" for "Jon Snow"
        And I complete the addressing step
        Then I should be on the checkout shipping step 
```
Same for selecting shipping and payment methods.

What is more, I think we should split the feature file into multiple features. E.g.
- Feature 1: Readdressing order
- Feature 2: Reselecting shipping method
- Feature 3: Reselecting payment method
- Feature 4: Placing order after moving back to the previous steps of the checkout

 WDYT?

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php (hunk 4 above)
Comment (Arminek) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: I thought that we have left this convention.

**Comment by michalmarcinkowski** [comment]
Great job Arek! 👍  Please split the feature file in a separate PR.

**Pull request merged**